server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java

```diff
@@ -152,6 +169,174 @@ public DenseVectorFieldMapper build(MapperBuilderContext context) {
+
+        INT8(1) {
+
+            @Override
+            public String toString() {
+                return "byte";
+            }
+
+            @Override
+            KnnVectorField createKnnVectorField(String name, float[] vector, VectorSimilarityFunction function) {
+                byte[] bytes = new byte[vector.length];
+
+                int index = 0;
+                while (index < vector.length) {
+                    bytes[index] = (byte) vector[index++];
+                }
+
+                return new KnnVectorField(name, new BytesRef(bytes), function);
+            }
+
+            @Override
+            IndexFieldData.Builder fielddataBuilder(DenseVectorFieldType denseVectorFieldType, FieldDataContext fieldDataContext) {
+                throw new IllegalArgumentException(
+                    "Fielddata is not supported on field ["
+                        + name()
+                        + "] of type ["
+                        + denseVectorFieldType.typeName()
+                        + "] "
+                        + "with element_type ["
+                        + this
+                        + "]"
+                );
+            }
+
+            @Override
+            void checkVectorBounds(float[] vector) {
+                StringBuilder errorBuilder = null;
+
+                for (int index = 0; index < vector.length; ++index) {
+                    float value = vector[index];
+                    double floor = Math.floor(value);
+
+                    if (value - floor > 1e-4 || value < Byte.MIN_VALUE || value > Byte.MAX_VALUE) {
+                        errorBuilder = new StringBuilder(
+                            this
+                                + " element_type vectors only support integers between"
+                                + "["
+                                + Byte.MIN_VALUE
+                                + ", "
+                                + Byte.MAX_VALUE
+                                + "] but found ["
+                                + value
+                                + "] at index ["
+                                + index
+                                + "];"
+                        );
+                        break;
+                    }
+                }
+
+                if (errorBuilder != null) {
+                    throw new IllegalArgumentException(appendErrorElements(errorBuilder, vector).toString());
+                }
+            }
+
+            @Override
+            void checkVectorMagnitude(VectorSimilarity similarity, float[] vector, float squaredMagnitude) {
+                StringBuilder errorBuilder = null;
+
+                if (similarity == VectorSimilarity.cosine && Math.sqrt(squaredMagnitude) == 0.0f) {
+                    errorBuilder = new StringBuilder(
+                        "The [" + VectorSimilarity.cosine.name() + "] similarity does not support vectors with zero magnitude."
+                    );
+                }
```

Comment:
Are we going to allow `dot_product` as well?

It stinks that we cannot provide the same quality checks here to determine that all vectors have similar magnitude. `int8` vectors with wildly different magnitudes will behave strangely with `dot_product`.

Reply:
This was a suggestion from @jtibshirani that we don't want to disallow dot_product under the circumstance where the user does ensure their vectors all have the same magnitude. I agree that it's unfortunate that we can't really protect this. 

One possible option would be to add an additional (optional?) parameter where the user let's us know the expected magnitude ahead of time, but that's pretty clunky. I'm not sure how useful this is, but if we did that maybe we could extend it to float types as well instead of requiring a magnitude of 1.0?

Reply:
It is debatable if having the fixed magnitude is worth it in document scoring. If we had the expected magnitude, and the quantization was done on unit-vectors, the vector distance calculations would have the same exact distances. 

Otherwise we end up getting different distances (according to that magnitude scale), but what is "close and far" end up being the same.

For consistency's sake, I would personally like a magnitude (that then gets passed down to lucene for the distance calculation), but that decision requires more thought and folks behind it.

Reply:
This is an important question, maybe we can discuss together "offline"? Also, once we decide we should update the `dense-vector-similarity` section in `dense-vector.asciidoc`. Currently it gives formulas for how the score is calculated, like `(1 + dot_product(query, vector)) / 2`, but these are different for byte-sized vectors.

Reply:
I added a few lines to the documentation on the difference for `dot_product` similarity between `float` and `byte`.

Reply:
When the `element_type` is `byte`, the score is actually different! You can take a look at Lucene's `VectorUtil#dotProductScore` to see the calculation.

Reply:
I see! Thank for pointing this out. I updated the docs to reflect the appropriate scoring function.